scipy/sparse/linalg/eigen/arpack/arpack.py

```diff
@@ -1790,40 +1800,60 @@ def svds(A, k=6, ncv=None, tol=0, which='LM', v0=None,
             X_dot = A.matvec
             X_matmat = A.matmat
             XH_dot = A.rmatvec
+            XH_mat = A.rmatmat
         else:
             X_dot = A.rmatvec
+            X_matmat = A.rmatmat
             XH_dot = A.matvec
+            XH_mat = A.matmat
 
             dtype = getattr(A, 'dtype', None)
             if dtype is None:
-                dtype = A.dot(np.zeros([m,1])).dtype
-
-            # A^H * V; works around lack of LinearOperator.adjoint.
-            # XXX This can be slow!
-            def X_matmat(V):
-                out = np.empty((V.shape[1], m), dtype=dtype)
-                for i, col in enumerate(V.T):
-                    out[i, :] = A.rmatvec(col.reshape(-1, 1)).T
-                return out.T
+                dtype = A.dot(np.zeros([m, 1])).dtype
```

Comment:
Do we have to perform this full dot product here? If `m` is, say, 1e5 it is certainly a waste just to get the `dtype`. We can use a single element of `A` and 0. which is `float` for the result type. But if this is only to test complex or real we can directly test for the complex dtype.

Reply:
If you reach here (object with no/None dtype attribute), what you have is a LinearOperator with unknown dtype. It's not possible to extract an element from it, except via matvec.

Reply:
I agree but we are possibly upcasting to either `float` or `complex` in the end anyways because of `np.zeros` being `float`. so there is a binary choice. A less computation intensive option (asking specifically for dtype etc.) might serve better here.

Reply:
The cost is one more matvec, whereas ARPACK/lobpcg will likely perform quite many of those.

What do you propose as an alternative? Raise error? That's going to be backward compatibility break, and I think the cost here (corner-case for users who did not explicitly specify dtype for their linear operator) is not large enough for that.

Reply:
OK I can't judge how many matvecs are used in a typical go so if it is negligible then I am OK with it.

Reply:
```
import numpy as np
from scipy.sparse.linalg import svds, LinearOperator
from scipy.sparse import diags

A = diags([np.arange(1, 10)**2], [0])
count = 0

def matvec(x):
    global count
    count += 1
    return A.dot(x)

def rmatvec(x):
    global count
    count += 1
    return A.T.dot(x)

op = LinearOperator(matvec=matvec, rmatvec=rmatvec, shape=A.shape)
res = svds(op, k=1, tol=1e-3, return_singular_vectors=False)
print("matvecs:", count)
```
prints 19.

Reply:
Nice. Then my concern is not relevant indeed.

Reply:
@ilayn - thanks for the review! 
@pv thanks for resolving the issue!